blobstore-uploader/src/main/java/org/jclouds/examples/blobstore/BlobUploader.java

```diff
@@ -0,0 +1,156 @@
+    * the command line.
+    */
+   private static ThreadLocal<BlobStore> blobStore = new ThreadLocal<BlobStore>();
+   private static ThreadLocal<String> container = new ThreadLocal<String>();
+
+   /**
+    * The only thing that really needs to be passed to this unit of work is the File to be uploaded.
+    * The credentials are here for convenience.
+    */
+   private String username;
+   private String password;
+   private String provider;
+   private String region;
+   private File file;
+
+   public
+   BlobUploader(String username, String password, String provider, String region, File file) {
+      this.username = username;
+      this.password = password;
+      this.provider = provider;
+      this.region = region;
+      this.file = file;
+   }
+
+   @Override
+   public void run() {
+      /**
+       * Instantiate the ThreadLocal variables when this thread runs for the first time.
+       * Instantiating this in the constructor will not work (different thread).
+       */
+      if(blobStore.get() == null) {
+         // It is usually a good idea to include the currentThread when logging parallel tasks.
+         System.out.println("Creating connection for thread " + Thread.currentThread());
+         /**
+          * In some cases, especially when running very large jobs with many parallel threads, some connections will
+          * break. In that case, we need to be able to obtain a new connection (and socket) to the service, which is
+          * why this is factored out.
+          */
+         resetBlobstore(username, password, provider, region);
+      }
+
+      if(container.get() == null) {
+         container.set(UUID.randomUUID().toString());
+         Location location = getOnlyElement(blobStore.get().listAssignableLocations());
+         blobStore.get().createContainerInLocation(location, container.get());
+
+         System.out.println("Created container " + container.get() +
+               " for thread " + Thread.currentThread() +
+               " in " + location.toString());
+      }
+
+      // The md5 as returned by the service, and as calculated locally.
+      String md5Local, md5Remote;
+      Blob blob = null;
+      FileInputStream fis = null;
+      try {
+         fis = new FileInputStream(file);
+         md5Local = DigestUtils.md5Hex(fis);
+         fis.close();
+      } catch (java.io.IOException e) {
+         e.printStackTrace();
+         /**
+          * The file is no longer available on the local FS.
+          * In some application cases, you might also want to retry this instead of finishing the unit of work.
+          */
+         return;
+      }
+
+         /**
+          * Uploading a file over a network is an inherently fragile operation. Over thousands of files, especially in
+          * highly parallel jobs that tax upload bandwidth, a small percent of uploads are guaranteed to fail.
+          */

```

Comment:
The retry loop is just for md5 mismatches.... will the ByteSourcePayload handle that? Then I won't need to calculate md5 at all. But I don't think it actually does that...?